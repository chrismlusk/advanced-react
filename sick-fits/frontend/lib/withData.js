import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

// this function creates our Apollo store

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    // local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // 1. read the `cartOpen` value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY
            });

            // 2. set cart state to the opposite value
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
