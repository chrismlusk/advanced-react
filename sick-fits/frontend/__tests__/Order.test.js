import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { fakeOrder } from '../lib/testUtils';

const testOrder = fakeOrder();

const mocks = [
  {
    request: { query: SINGLE_ORDER_QUERY, variables: { id: testOrder.id } },
    result: { data: { order: fakeOrder() } }
  }
];

describe('<Order />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Order id={testOrder.id} />
      </MockedProvider>
    );
  });

  it('renders and matches the snapshot', async () => {
    await wait();
    wrapper.update();
    const order = wrapper.find('div[data-test="order"]');
    const OrderItem = wrapper.find('OrderItem');
    expect(toJSON(order)).toMatchSnapshot();
    expect(OrderItem).toHaveLength(2);
  });
});
