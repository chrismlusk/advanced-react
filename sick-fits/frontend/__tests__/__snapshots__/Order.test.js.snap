// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders and matches the snapshot 1`] = `
<div
  className="OrderStyles-sc-4oqalm-0 camMij"
  data-test="order"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <p>
    <span>
      Order ID
    </span>
    <span>
      ord123
    </span>
  </p>
  <p>
    <span>
      Charge
    </span>
    <span>
      ch_123
    </span>
  </p>
  <p>
    <span>
      Date
    </span>
    <span>
      March 31, 2018 7:00 PM
    </span>
  </p>
  <p>
    <span>
      Total
    </span>
    <span>
      $400
    </span>
  </p>
  <p>
    <span>
      Item Count
    </span>
    <span>
      2
    </span>
  </p>
  <div
    className="items"
  >
    <OrderItem
      item={
        Object {
          "__typename": "OrderItem",
          "description": "expedita est doloremque explicabo eaque necessitatibus sit",
          "id": "b4505a74-1f67-4ee1-9558-e56760116bcd",
          "image": "dolorem.jpg",
          "price": 4234,
          "quantity": 1,
          "title": "nihil quam nulla est unde sit ex",
        }
      }
      key="b4505a74-1f67-4ee1-9558-e56760116bcd"
    >
      <div
        className="order-item"
      >
        <img
          alt="nihil quam nulla est unde sit ex"
          src="dolorem.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            nihil quam nulla est unde sit ex
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
          <p>
            expedita est doloremque explicabo eaque necessitatibus sit
          </p>
        </div>
      </div>
    </OrderItem>
    <OrderItem
      item={
        Object {
          "__typename": "OrderItem",
          "description": "quis quae iusto sapiente nostrum aut corrupti",
          "id": "4cef4042-dadc-4ad4-a066-08d01ced82d4",
          "image": "et.jpg",
          "price": 4234,
          "quantity": 1,
          "title": "voluptas culpa et vero repellat eveniet assumenda",
        }
      }
      key="4cef4042-dadc-4ad4-a066-08d01ced82d4"
    >
      <div
        className="order-item"
      >
        <img
          alt="voluptas culpa et vero repellat eveniet assumenda"
          src="et.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            voluptas culpa et vero repellat eveniet assumenda
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
          <p>
            quis quae iusto sapiente nostrum aut corrupti
          </p>
        </div>
      </div>
    </OrderItem>
  </div>
</div>
`;
